schema/representations.md

```diff
@@ -0,0 +1,256 @@
+
+```ipldsch
+type TypeList struct {
+	valueType TypeTerm
+	valueNullable Bool
+} representation map {
+	field valueNullable default "false"
+}
+```
+
+#### struct tuple representation example schema
+
+```ipldsch
+```
+
+#### struct tuple representation example data
+
+```json
+```
+
+### struct stringjoin representation example schema
+
+```ipldsch
+## Fizzlebop is a pair of fields which serializes as "part1:part2" as a string.
+type Fizzlebop struct {
+	a String
+	b String
+} representation stringjoin {
+	join ":"
+}
+```
+
+### struct stringpairs representation example schema
+
+Say we're doing something awfully like the mount options in an /etc/fstab file:
+
+```ipldsch
+## MountOptions serializes as a string of comma-separated "k=v" entries, e.g. "k1=v1,k2=v2".
+## This serial representation is limited (n.b. The domain of keys must exclude "=" and values must exclude ","!),
+## but we've chosen it because other parts of our application protocol are drastically simplified by limiting this to a string.
+type MountOptions map {String:String} representation stringpairs {
+	innerDelim "="
+	entryDelim ","
+}
+```
+
+### union keyed representation example
+
+```ipldsch
+type MyKeyedUnion union {
+	| Foo "foo"
+	| Bar "bar"
+} representation keyed
+
+type Foo struct { froz Bool }
+type Bar int
+```
+
+Some data matching `MyKeyedUnion` (shown as JSON) is:
+
+```json
+{"foo": {"froz": true}}
+```
+
+This data would also match, as the other type:
+
+```json
+{"bar": 12}
+```
+
+Note how kinded unions don't introduce any kind of wrapping object at all;
+contrast this with each of the other union representation strategies, all
+of which use at least one layer of map in their representation.
+
+### union kinded representation example
+
+```ipldsch
+type MyKindedUnion union {
+	| Foo map
+	| Bar int
+} representation kinded
+
+type Foo struct { froz Bool }
+type Bar int
+```
+
+Some data matching `MyKindedUnion` (shown as JSON) is:
+
+```json
+{"froz": true}
+```
+
+This data would also match, as the other type:
+
+```json
+12
+```
+
+Note that the syntax used in the type declaration is different for kinded
+unions versus other union representations!
+Kinded unions, unlike all other representation strategies, don't have
+any kind of string keywords at all.  Instead, they encode all of their
+discriminant information in the kind of the data itself!
+Correspondingly, the tagging information in the type definition uses kind
+keywords (unquoted) instead of quoted strings.
+
+Note that the kinds valid in a kinded union are all Data Model-layer
+*representation kinds*.  E.g. `string` and `map` are acceptable here;
+`struct` is not.  The kind that a union member type is tagged with must
+match that type's representation kind or the schema is invalid.
+
+### union envelope representation example
+
+```ipldsch
+type MyEnvelopeUnion union {
+	| Foo "foo"
+	| Bar "bar"
+} representation kinded {
+	discriminantKey "tag"
```

Comment:
does it make sense to have default keys for envelopes? `type` and `value` maybe?

Reply:
I'm not sure.

I think I'd vote against because it introduce more opinions into the core; and reduces the amount of strings which are can be expected to show up word-for-word in both the schema and the encoded data, which I think are a nice sanity synchronizer.  It's a weak vote, though.

(Also, I'll readily admit "discriminantKey" is heckin' long and a handful to type.  Maybe something terser here, or using positional arguments for this, would improve the feel.)